kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportMapper.kt

```diff
@@ -75,8 +75,15 @@ private fun isSealedClassConstructor(descriptor: ConstructorDescriptor) = descri
 
 // Note: partially duplicated in ObjCExportLazyImpl.translateTopLevels.
 internal fun ObjCExportMapper.shouldBeExposed(descriptor: CallableMemberDescriptor): Boolean =
-        descriptor.isEffectivelyPublicApi && !descriptor.isExpect &&
-                !isHiddenByDeprecation(descriptor) && !(descriptor is ConstructorDescriptor && isSealedClassConstructor(descriptor))
+        descriptor.isEffectivelyPublicApi && !descriptor.isExpect && !isHiddenByDeprecation(descriptor) &&
+                !(descriptor is ConstructorDescriptor && isSealedClassConstructor(descriptor)) && !descriptor.isObjCRefined()
+
+private fun CallableMemberDescriptor.isObjCRefined(): Boolean {
+    if (overriddenDescriptors.isNotEmpty()) return overriddenDescriptors.all { it.isObjCRefined() }
+    return annotations.any { it.fqName == KonanFqNames.objCRefined } || annotations.any { annotation ->
+        annotation.annotationClass?.annotations?.any { it.fqName == KonanFqNames.objCRefined } == true
```

Comment:
This line enables specific behaviour. I have a number of concerns regarding this.

1. This behaviour is not documented.
2. What is it required for? What is the use case?
3. Kotlin has other examples of similar meta-annotation approach (e.g. `RequiresOptIn`), but I can't remember ones where annotation works in both cases -- when applying it directly and when applying an annotation annotated with it. So, this might be a bit inconsistent with other Kotlin features.
4. Should this work when the chain is longer? E.g.
    ```kotlin
    @ObjCRefined
    annotation class MyObjCRefined

    @MyObjCRefined
    annotation class MyObjCRefined2

    @MyObjCRefined2
    val amIExported = 42
    ```

Reply:
> What is it required for? What is the use case?

The use case is to support annotation processors. This would allow a processor to define its own annotation (e.g. `GenerateCoolObjCAPI`) that inherits the `ObjCRefined` (or `SwiftRefined`) annotation.
This would prevent users from having to apply both annotations.

> Kotlin has other examples of similar meta-annotation approach (e.g. RequiresOptIn), but I can't remember ones where annotation works in both cases -- when applying it directly and when applying an annotation annotated with it. So, this might be a bit inconsistent with other Kotlin features.

Would it be better to have a separate meta-annotation? How should we call that?

> Should this work when the chain is longer?

No that shouldn't work with the current proposal.
Do you have a specific use case for that?

Reply:
> The use case is to support annotation processors. This would allow a processor to define its own annotation (e.g. `GenerateCoolObjCAPI`) that inherits the `ObjCRefined` (or `SwiftRefined`) annotation.
This would prevent users from having to apply both annotations.

Ok. And why would the processor apply its own annotation in addition to `ObjCRefined` or `SwiftRefined`? Could you please provide an example?

> Would it be better to have a separate meta-annotation? How should we call that?

Yes, this would be better than the current version. Not sure about the name yet. Please take a look at other annotations that are applicable to annotations, this might give a hint.

> No that shouldn't work with the current proposal.
Do you have a specific use case for that?

I don't have a specific case, but this is just a bit inconsistent. Like "applying `@ObjCRefined` to an annotation makes the latter behave like `@ObjCRefined` itself, but not quite". So we already have two different entities -- `@ObjCRefined` and "almost `@ObjCRefined`".
Introducing a separate meta annotation would of course remove the concern.

Reply:
> Ok. And why would the processor apply its own annotation in addition to `ObjCRefined` or `SwiftRefined`? Could you please provide an example?

Say I want to expose `Int`s as `String`s (or all `Flow`s as callbacks). I could then create an annotation processor that generates the expected declarations for me while automatically hiding the original.
```kotlin
@IntAsString
fun returnSomeInt(): Int = TODO()

// auto-generated by the annotation processor
fun returnSomeIntAsString(): String = TODO() 
```
Allow the `IntAsString` annotation to "inherit" the `ObjCRefined` annotation would then hide `returnSomeInt` from the ObjC API. Combined with the `ObjCName` annotation (from #4815) the generated function could even have the ObjC name of the original function. This would allow annotation processors to improve the ObjC and Swift API automatically (without having "duplicate" declarations in your public API).

> Introducing a separate meta annotation would of course remove the concern.
Agreed! Having a separate meta annotation would be easier to understand.

Reply:
Thank you for the explanation!

> Say I want to expose `Int`s as `String`s (or all `Flow`s as callbacks). I could then create an annotation processor that generates the expected declarations for me while automatically hiding the original.

Is this annotation process a compiler plugin? Can't it simply add `@ObjCRefined` to `returnSomeInt`?

Reply:
> Is this annotation process a compiler plugin? Can't it simply add `@ObjCRefined` to `returnSomeInt`?

Can compiler plugins add annotations? That would be cool, but that would still require a full compiler plugin.
With the meta-annotation a KSP plugin could be used as well (which is more maintainable).

Reply:
> Can compiler plugins add annotations?

Yes.

> That would be cool, but that would still require a full compiler plugin.
With the meta-annotation a KSP plugin could be used as well (which is more maintainable).

Do you mean that a KSP plugin can't add an annotation to the existing declaration?
Can it create a new declaration then?

Reply:
> Do you mean that a KSP plugin can't add an annotation to the existing declaration?
> Can it create a new declaration then?

Yeah KSP plugins can't modify existing code. They can only generate additional code.
So it can't create a new declaration inside the existing classes, but it can create extension properties/functions.
Which (in most cases) are exported as class members to ObjC/Swift.

Reply:
Ah, ok.
Fair enough, let's have annotations and meta annotations.

Reply:
We now have annotations and meta annotations. The export logic will only need to check for the meta annotations since the built in annotations are also annotated with the meta annotations.

It's still not allowed to chain the annotations. The compiler should enforce this by only allowing the meta annotations on annotation classes with the property and/or function targets.